pkg/gofr/datasource/file/ftp/fs.go

```diff
@@ -0,0 +1,353 @@
+
+			dirs = append([]string{dir}, dirs...)
+			if parentDir == "" || parentDir == "/" {
+				break
+			}
+			currentdir = path.Clean(parentDir)
+		} else {
+			dirs = append([]string{currentdir}, dirs...)
+			break
+		}
+
+	}
+
+	return dirs
+}
+
+// MkdirAll creates directories recursively on the FTP server. Here, os.FileMode is unused.
+// The directories are not created if even one directory exist.
+func (f *ftpFileSystem) MkdirAll(name string, _ os.FileMode) error {
+	defer f.processLog(&FileLog{Operation: "MkdirAll", Location: path.Join(f.config.RemoteDir, name)}, time.Now())
+
+	if name == "" {
+		f.logger.Errorf("MkdirAll failed. Provide a valid path : %v", errEmptyPath)
+		return errEmptyPath
+	}
+	// returns a slice of those directories that do not exist with the first index being the latest existing parent directory path.
+	dirs := f.mkdirAllHelper(name)
+
+	currentDir := dirs[0]
+
+	for i, dir := range dirs {
+		if i == 0 {
+			continue
+		} else {
+			currentDir = path.Join(currentDir, dir)
+		}
+
+		err := f.conn.MakeDir(currentDir)
+		if err != nil {
+			f.logger.Errorf("MkdirAll failed : %v", err)
+
+			return err
+		}
+	}
+
+	f.logger.Logf("Directories creation completed successfully.")
+
+	return nil
+}
+
+// Open retrieves a file from the FTP server and returns a file handle.
+func (f *ftpFileSystem) Open(name string) (File, error) {
+	filePath := path.Join(f.config.RemoteDir, name)
+
+	defer f.processLog(&FileLog{Operation: "Open", Location: filePath}, time.Now())
+
+	if name == "" {
+		f.logger.Errorf("Open_file failed. Provide a valid filename : %v", errEmptyFilename)
+		return nil, errEmptyFilename
+	}
+
+	res, err := f.conn.Retr(filePath)
+	if err != nil {
+		f.logger.Errorf("Open_file failed. Error opening file : %v", err)
+		return nil, err
+	}
+
+	_, s := path.Split(filePath)
+
+	f.logger.Logf("Open_file successful. Filepath : %q", filePath)
+
+	defer res.Close()
+
+	return &ftpFile{
+		response: res,
+		name:     s,
+		path:     filePath,
+		conn:     f.conn,
+		logger:   f.logger,
+		metrics:  f.metrics,
+	}, nil
+}
+
+// permissions are not clear for Ftp as file commands do not accept an argument and don't store their file permissions.
+// currently, this function just calls the Open function. Here, os.FileMode is unused.
+func (f *ftpFileSystem) OpenFile(name string, _ int, _ os.FileMode) (File, error) {
+	return f.Open(name)
+}
+
+// Remove deletes a file from the FTP server.
+func (f *ftpFileSystem) Remove(name string) error {
+	filePath := path.Join(f.config.RemoteDir, name)
+
+	defer f.processLog(&FileLog{Operation: "Remove", Location: filePath}, time.Now())
+
+	if name == "" {
+		f.logger.Errorf("Remove_file failed. Provide a valid filename : %v", errEmptyFilename)
+		return errEmptyFilename
+	}
+
+	err := f.conn.Delete(filePath)
+
+	if err != nil {
+		var textprotoError *textproto.Error
+
+		if errors.As(err, &textprotoError) && textprotoError.Msg == transferCompleteError.Msg {
```

Comment:
```suggestion
		if errors.As(err, &textprotoError) && textprotoError.Code == errTransferComplete.Code {
```

I would have checked only the code

Reply:
But then

I would have used this

```suggestion
		if errors.As(err, &textprotoError) && textprotoError.Code == textproto.StatusClosingDataConnection {
```